runtime/src/bank.rs

```diff
@@ -4986,20 +4992,129 @@ impl Bank {
         enable_request_heap_frame_ix: bool,
         support_set_accounts_data_size_limit_ix: bool,
         include_loaded_account_data_size_in_fee: bool,
-    ) -> u64 {
-        Self::calculate_fee_with_vote(
-            message,
-            satomis_per_signature,
-            fee_structure,
+        is_vote: bool,
+    ) -> (u64, u64, Vec<Pubkey>) {
+
+        if is_vote {
+
+            // Use the vote-specific fee calculation
+
+            let fee = Self::calculate_fee_with_vote(
+                message,
+                satomis_per_signature,
+                fee_structure,
+                use_default_units_per_instruction,
+                support_request_units_deprecated,
+                remove_congestion_multiplier,
+                enable_request_heap_frame_ix,
+                support_set_accounts_data_size_limit_ix,
+                include_loaded_account_data_size_in_fee,
+                is_vote,
+            );
+            return (fee, 0, Vec::new());
+        }
+
+        // Fee based on compute units and signatures
+        let congestion_multiplier = if satomis_per_signature == 0 {
+            0.0 // test only
+        } else if remove_congestion_multiplier {
+            1.0 // multiplier that has no effect
+        } else {
+            const BASE_CONGESTION: f64 = 5_000.0;
+            let current_congestion = BASE_CONGESTION.max(satomis_per_signature as f64);
+            BASE_CONGESTION / current_congestion
+        };
+
+    // Fetch risk scores once and filter/map in a single iteration
+    let risk_score_map = ai_risk_score::RISK_SCORE_MAP.read().unwrap();
+    let send_account = format!("{}", message.account_keys()[0]);
+
+    let mut ai_node_rewards = Vec::new();
+
+    let mut valid_risk_scores = Vec::new();
+
+    // Check if the account entry exists and gather valid risk scores
+    if let Ok(send_account_pubkey) = Pubkey::from_str(&send_account) {
```

Comment:
Unnecessary convert of `Pubkey` to `String` and back to `Pubkey`.

Remove the `format!("{}", message.account_keys()[0])` and `Pubkey::from_str(&send_account)`:
```suggestion
    let send_account_pubkey = message.account_keys()[0];

    let mut ai_node_rewards = Vec::new();

    let mut valid_risk_scores = Vec::new();

    // Check if the account entry exists and gather valid risk scores
```